quota/etcd/storage/quota_storage.go

```diff
@@ -0,0 +1,387 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package storage contains storage classes for etcd-based quotas.
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian/quota"
+	"github.com/google/trillian/quota/etcd/storagepb"
+)
+
+const (
+	configsKey = "quotas/configs"
+)
+
+var (
+	// timeNow is used in place of time.Now, so it may be replaced by tests.
+	timeNow = time.Now
+
+	globalPattern *regexp.Regexp
+	treesPattern  *regexp.Regexp
+	usersPattern  *regexp.Regexp
+)
+
+func init() {
+	var err error
+	globalPattern, err = regexp.Compile("quotas/global/(read|write)/config")
+	if err != nil {
+		glog.Fatalf("bad global pattern: %v", err)
+	}
+	treesPattern, err = regexp.Compile("quotas/trees/\\d+/(read|write)/config")
+	if err != nil {
+		glog.Fatalf("bad trees pattern: %v", err)
+	}
+	usersPattern, err = regexp.Compile("quotas/users/[^/]+/(read|write)/config")
+	if err != nil {
+		glog.Fatalf("bad users pattern: %v", err)
+	}
+}
+
+// QuotaStorage is the interface between the etcd-based quota implementations (quota.Manager and
+// RPCs) and etcd itself.
+type QuotaStorage struct {
+	Client *clientv3.Client
+}
+
+// UpdateConfigs creates or updates the supplied configs in etcd.
+// If no config exists, the current config is assumed to be an empty storagepb.Configs proto.
+// The update function allows for mask-based updates and ensures a single-transaction
+// read-modify-write operation.
+// If reset is true, all specified configs will be set to their max number of tokens. If false,
+// existing quotas won't be modified, unless the max number of tokens is downgraded, in which case
+// the new ceiling is enforced.
+// Newly created quotas are always set to max tokens, regardless of the reset parameter.
+func (qs *QuotaStorage) UpdateConfigs(ctx context.Context, reset bool, update func(*storagepb.Configs)) (*storagepb.Configs, error) {
+	if update == nil {
+		return nil, errors.New("update function required")
+	}
+
+	var updated *storagepb.Configs
+	_, err := concurrency.NewSTMSerializable(ctx, qs.Client, func(s concurrency.STM) error {
+		previous, err := getConfigs(s)
+		if err != nil {
+			return err
+		}
+		// Take a deep copy of "previous". It's pointers all the way down, so it's easier to just
+		// unmarshal it again. STM has the key we just read and it should be exactly the same as
+		// previous, so it's ok to ignore the error...
```

Comment:
Since we sanity check (and potentially return an error) below, I think it would only add noise. I don't really expect this to happen, I'm mostly playing it safe here.